### DISCUSSION TITLE
[ZK Stack] - withdraw finalize keeps pending on dockerized hyperchain env when using cpu prover

### DISCUSSION THREAD

**Discussion by tiagorvmartins** [zkSync-Community-Hub/zksync-developers#293] [ZK Stack] [answered]
### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [ ] Server (zksync-era)
- [X] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

nightly-2023-08-21-x86_64-unknown-linux-gnu

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

Currently and following the documentation, having my own hyperchain based on docker setup command and using prover setup (CPU) doesn't work out of the box without several changes.

Here are the entire changes we did in order to have the environment up and running:

1) git clone of repo zksync-era on tag core-v20.1.0

2) created the volumes folder with the current user (since the wizard was creating with root ownership) and then failing further in the process.

3) removed the 127.0.0.1 from docker compose stack common, because the wizard wasn't able to reach the network geth localhost with that binding.

4) updated the library in package.json inside contracts/system-contracts 
from:
      "@matterlabs/hardhat-zksync-solc": "^0.4.2",
to
      "@matterlabs/hardhat-zksync-solc": "^1.0.6",
Because the wizard was failing with undefined dependency when trying to use zsolc.

5) changed the docker image being used on the compose file (services: zkstack-core, zkstack-apis) that is generated for our hyperchain environment (when using zk stack docker-setup)

from:
      our-env/server-v2:latest
to:
      matterlabs/server-v2:our-env-linux-amd64

6) Changed our our-env.env

from "localhost" hostname to "geth" on ETH_CLIENT_WEB3_URL 
from "localhost" hostname to "postgres" on DATABASE_URL and TEST_DATABASE_URL

Then after the wizard complete with success and opting for not spinning up the server, we did execute zk stack prover-setup choosing CPU prover and downloading the keys.
And finally we generated the compose file using zk stack docker-setup, where we required the change on 5) and 6) to properly launch the hyperchain without failing.

But then when doing deposits/withdraws it fails on withdraw-finalize.
Below is the list of commands executed and their outputs.

**Command 1)**
npx zksync-cli@latest wallet balance --chain dockerized-node --address 0xa61464658AfeAf65CccaaFD3a512b69A83B77618
**Result 1)**
Dockerized local node Balance: 0.0026064036125 ETH

**Command 2)**
npx zksync-cli@latest bridge deposit
**Result 2)**
Deposit:
 From: 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049 (L1 Local)
 To: 0xa61464658AfeAf65CccaaFD3a512b69A83B77618 (Dockerized local node)
 Amount: 10 ETH
Sending deposit transaction...
Deposit sent:
 Transaction hash: 0xc09eb64f0a5c391bf7849a6a3a43fb51d729ff21f21a329d4bec7233f9beffcc
Sender L1 balance after transaction: 99999999999999999999.873483478541131798 ETH

**Command 3)**
npx zksync-cli@latest wallet balance --chain dockerized-node --address 0xa61464658AfeAf65CccaaFD3a512b69A83B77618
**Result 3)**
dockerized local node Balance: 10.0026064036125 ETH

**Command 4)**
npx zksync-cli@latest bridge withdraw --amount 2 --chain dockerized-node --to 0xc0ffee254729296a45a3885639AC7E10F9d54979 --pk 0xac1e735be8536c65
34bb4f17f06f6afc73b2b5ba84ac2cfb12f7461b20c0bbe3
**Result 4)**
Withdraw:
 From: 0xa61464658AfeAf65CccaaFD3a512b69A83B77618 (Dockerized local node)
 To: 0xc0ffee254729296a45a3885639AC7E10F9d54979 (L1 Local)
 Amount: 2 ETH
Sending withdraw transaction...
Withdraw sent:
 Transaction hash: 0x709cbf32345797b08773f3c3890a787753afdf117e4b29ea79251e65f290999f
Sender L2 balance after transaction: 10.0026064036125 ETH

**Command 5:**
npx zksync-cli@latest wallet balance --chain dockerized-node --address 0xa61464658AfeAf65CccaaFD3a512b69A83B77618
**Result 5:**
Dockerized local node Balance: 8.0025860392125 ETH

**Command 6:**
npx zksync-cli@latest bridge withdraw-finalize --hash 0x709cbf32345797b08773f3c3890a787753afdf117e4b29ea79251e65f290999f --pk 0xac1e735be8536c6534bb4f17f06f6afc73b2b5ba84ac2cfb12f7461b20c0bbe3 --chain dockerized-node
**Result 6:**
Withdraw finalize:
 From chain: Dockerized local node
 To chain: L1 Local
 Withdrawal transaction (L2): 0x709cbf32345797b08773f3c3890a787753afdf117e4b29ea79251e65f290999f
 Finalizer address (L1): 0xa61464658AfeAf65CccaaFD3a512b69A83B77618
Checking status of the transaction...
ⓘ 
Transaction is still being processed on Dockerized local node, please try again when the ethExecuteTxHash has been computed
L2 Transaction Details: {
  "isL1Originated": false,
  "status": "included",
  "fee": "0x128574d3fc00",
  "gasPerPubdata": "0xc350",
  "initiatorAddress": "0xa61464658afeaf65cccaafd3a512b69a83b77618",
  "receivedAt": "2024-01-25T14:23:52.055Z",
  "ethCommitTxHash": "0xb87bb305fef8329c4600872f213b9a2f1b1a8e78255151d642db099c4e794811",
  "ethProveTxHash": null,
  "ethExecuteTxHash": null
}

**Command 7)**
curl -X POST -H "Content-Type: application/json" \
--data '{"jsonrpc": "2.0", "id": 1, "method": "zks_getTransactionDetails", "params": [ "0x709cbf32345797b08773f3c3890a787753afdf117e4b29ea79251e65f290999f" ]}' \
"http://localhost:3050"
**Result 7)**
{"jsonrpc":"2.0","result":{"isL1Originated":false,"status":"included","fee":"0x128574d3fc00","gasPerPubdata":"0xc350","initiatorAddress":"0xa61464658afeaf65cccaafd3a512b69a83b77618","receivedAt":"2024-01-25T14:23:52.055Z","ethCommitTxHash":"0xb87bb305fef8329c4600872f213b9a2f1b1a8e78255151d642db099c4e794811","ethProveTxHash":null,"ethExecuteTxHash":null},"id":1}

### Expected Behavior

We were expecting the withdraw-finalize command to succeed and the tokens to get withdraw to L1 account.

### Current Behavior

Tokens are lost, when withdrawing from L2 -> L1 when using CPU Prover.

### Additional Details

Unless we are missing something, we would like to understand why this is happening.

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by zk-Lumi** [comment]
Thanks for reaching out @tiagorvmartins,

Sent to team to take a look at - stay tuned.

**Comment by shahar4** [comment]
Hey @tiagorvmartins! The error message says `please try again when the ethExecuteTxHash has been computed`. The meaning is: A withdrawal can only be finalized once a proof has been computed, verified by the L1, and the `Execute` tx has been called in the L1 contract. If you want to try the finalizer, it might be easier to run with fake proofs.

**Comment by varnerrobertv** [comment]
It seems like the issue might be related to the Prover setup or the configuration of the environment. Ensure that all configurations, especially related to Prover and Docker setup, are correctly set up and aligned with the documentation. Additionally, consider reaching out to the community or the project maintainers for further assistance and troubleshooting.

**Comment by quanteauphil** [comment]
The issue seems to stem from a failure during the withdrawal-finalize process when using the CPU Prover setup for withdrawing tokens from Layer 2 (L2) to Layer 1 (L1) in zkSync. Despite following the setup instructions and making several adjustments to configurations, the withdrawal-finalize command does not succeed, resulting in the loss of tokens.

Potential avenues for resolution include double-checking configuration settings, reviewing prover logs for error messages, debugging the withdrawal process step by step, seeking assistance from the zkSync community or development team, reviewing documentation, and examining relevant code repositories for potential issues.

Further investigation is needed to identify the root cause and implement a solution to ensure successful withdrawals from L2 to L1.

**Comment by mcgeesusan** [comment]
It seems there might be an issue with the withdraw-finalize command when using the CPU prover setup. Verify the configuration and ensure all steps in the setup process are correctly followed. Additionally, consider checking the logs for any errors or inconsistencies during the withdrawal process.

**Comment by iscxyz** [comment, marked as answer]
:wave: We've done a few fixes recently that should have resolved this issue. Please try the latest `main` and let me know if it works now. CC @tiagorvmartins

**Comment by enessosyalmedya** [comment]
thanks for your effort

**Comment by webgolfrelenromero** [comment]
It appears that there are several setup issues and inconsistencies in the deployment process and configuration of the zkSync environment, leading to failures in withdrawing tokens from L2 to L1 when using CPU Prover. Further investigation and debugging are needed to identify and resolve the root cause of this issue.
